File: lib/keyboards.js
import { listButtonLabel } from 'lib/render';

// Схема callback data:
//   home | list:<sort>:<page> | card:<id> | pay:<id> | charge:<id> | refund:<id>
//   price:<id> | hist:<id>:<page> | add | search | sortmenu | undo | undo_yes | undo_no
//   export | exp_students | exp_history | exp_backup | cancel | noop

export const PAGE_SIZE = 10;
export const SORTS = { name: 'name', bal: 'bal', due: 'due' };

function btn(text, data) {
  return { text, callback_data: data };
}

// Ряд пагинации: рисуем только реальные стрелки — Telegram отклоняет кнопки
// с пустым текстом (BUTTON_TEXT_EMPTY).
function paginationRow(prefix, page, totalPages) {
  if (totalPages <= 1) return null;
  const row = [];
  if (page > 0) row.push(btn('⬅️', `${prefix}:${page - 1}`));
  row.push(btn(`${page + 1}/${totalPages}`, 'noop'));
  if (page < totalPages - 1) row.push(btn('➡️', `${prefix}:${page + 1}`));
  return row;
}

// Главное меню: кнопка на каждого ученика + служебный ряд.
export function mainMenuKeyboard(students, sort, page, totalPages) {
  const rows = students.map((s) => [btn(listButtonLabel(s), `card:${s.id}`)]);
  const pagination = paginationRow(`list:${sort}`, page, totalPages);
  if (pagination) rows.push(pagination);
  rows.push([btn('➕ Добавить', 'add'), btn('🔍 Поиск', 'search'), btn('↕️ Сортировка', 'sortmenu')]);
  rows.push([btn('↩️ Отменить действие', 'undo'), btn('📤 Экспорт', 'export')]);
  return { inline_keyboard: rows };
}

export function cardKeyboard(id) {
  return {
    inline_keyboard: [
      [btn('➕ Внести оплату', `pay:${id}`)],
      [btn('➖ Списать урок', `charge:${id}`), btn('↩️ Вернуть урок', `refund:${id}`)],
      [btn('✏️ Изменить стоимость', `price:${id}`), btn('📜 История', `hist:${id}:0`)],
      [btn('⬅️ К списку', 'home')],
    ],
  };
}

export function historyKeyboard(id, page, totalPages) {
  const rows = [];
  const pagination = paginationRow(`hist:${id}`, page, totalPages);
  if (pagination) rows.push(pagination);
  rows.push([btn('⬅️ К карточке', `card:${id}`)]);
  return { inline_keyboard: rows };
}

export function sortMenuKeyboard() {
  return {
    inline_keyboard: [
      [btn('🔤 По имени', `list:${SORTS.name}:0`)],
      [btn('📚 По остатку занятий', `list:${SORTS.bal}:0`)],
      [btn('⏳ Скоро потребуется оплата', `list:${SORTS.due}:0`)],
      [btn('⬅️ Назад', 'home')],
    ],
  };
}

// В data зашит id операции, чтобы отменить именно то, что показано в диалоге,
// а не «последнее на момент клика».
export function undoConfirmKeyboard(opId) {
  return {
    inline_keyboard: [[btn('✅ Да, отменить', `undo_yes:${opId}`), btn('❌ Нет', 'undo_no')]],
  };
}

export function exportKeyboard() {
  return {
    inline_keyboard: [
      [btn('👥 Ученики (CSV)', 'exp_students')],
      [btn('📜 История операций (CSV)', 'exp_history')],
      [btn('💾 Полный бэкап', 'exp_backup')],
      [btn('⬅️ Назад', 'home')],
    ],
  };
}

// Кнопка отмены текущего текстового ввода (FSM).
export function cancelKeyboard() {
  return { inline_keyboard: [[btn('❌ Отмена', 'cancel')]] };
}

export function homeKeyboard() {
  return { inline_keyboard: [[btn('⬅️ К списку', 'home')]] };
}

File: selfhosted/bot/keyboards.py
"""Inline-клавиатуры и схема callback data.

Схема:
  home | list:<sort>:<page> | card:<id> | pay:<id> | charge:<id> | refund:<id>
  price:<id> | hist:<id>:<page> | add | search | sortmenu | undo
  undo_yes:<opId> | undo_no | export | exp_xlsx | exp_csv | cancel | noop
"""

from aiogram.types import InlineKeyboardButton, InlineKeyboardMarkup

from .render import list_button_label

PAGE_SIZE = 10
HISTORY_PAGE_SIZE = 10


def _btn(text: str, data: str) -> InlineKeyboardButton:
    return InlineKeyboardButton(text=text, callback_data=data)


def _pagination_row(prefix: str, page: int, total_pages: int) -> list[InlineKeyboardButton] | None:
    """Только реальные стрелки — Telegram отклоняет кнопки с пустым текстом."""
    if total_pages <= 1:
        return None
    row = []
    if page > 0:
        row.append(_btn("⬅️", f"{prefix}:{page - 1}"))
    row.append(_btn(f"{page + 1}/{total_pages}", "noop"))
    if page < total_pages - 1:
        row.append(_btn("➡️", f"{prefix}:{page + 1}"))
    return row


def main_menu_kb(students, sort: str, page: int, total_pages: int) -> InlineKeyboardMarkup:
    rows = [[_btn(list_button_label(s), f"card:{s.id}")] for s in students]
    pagination = _pagination_row(f"list:{sort}", page, total_pages)
    if pagination:
        rows.append(pagination)
    rows.append([_btn("➕ Добавить", "add"), _btn("🔍 Поиск", "search"), _btn("↕️ Сортировка", "sortmenu")])
    rows.append([_btn("↩️ Отменить действие", "undo"), _btn("📤 Экспорт", "export")])
    return InlineKeyboardMarkup(inline_keyboard=rows)


def card_kb(sid: int) -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        inline_keyboard=[
            [_btn("➕ Внести оплату", f"pay:{sid}")],
            [_btn("➖ Списать урок", f"charge:{sid}"), _btn("↩️ Вернуть урок", f"refund:{sid}")],
            [_btn("✏️ Изменить стоимость", f"price:{sid}"), _btn("📜 История", f"hist:{sid}:0")],
            [_btn("⬅️ К списку", "home")],
        ]
    )


def history_kb(sid: int, page: int, total_pages: int) -> InlineKeyboardMarkup:
    rows = []
    pagination = _pagination_row(f"hist:{sid}", page, total_pages)
    if pagination:
        rows.append(pagination)
    rows.append([_btn("⬅️ К карточке", f"card:{sid}")])
    return InlineKeyboardMarkup(inline_keyboard=rows)


def sort_menu_kb() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        inline_keyboard=[
            [_btn("🔤 По имени", "list:name:0")],
            [_btn("📚 По остатку занятий", "list:bal:0")],
            [_btn("⏳ Скоро потребуется оплата", "list:due:0")],
            [_btn("⬅️ Назад", "home")],
        ]
    )


def undo_confirm_kb(op_id: int) -> InlineKeyboardMarkup:
    # op_id в data — отменяем именно показанную операцию, а не «последнюю на момент клика».
    return InlineKeyboardMarkup(
        inline_keyboard=[[_btn("✅ Да, отменить", f"undo_yes:{op_id}"), _btn("❌ Нет", "undo_no")]]
    )


def export_kb() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        inline_keyboard=[
            [_btn("📊 Excel (.xlsx)", "exp_xlsx")],
            [_btn("📄 CSV (.csv)", "exp_csv")],
            [_btn("⬅️ Назад", "home")],
        ]
    )


def cancel_kb() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(inline_keyboard=[[_btn("❌ Отмена", "cancel")]])


def home_kb() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(inline_keyboard=[[_btn("⬅️ К списку", "home")]])
